File: PyNet/ga2230Net/ListenerThread.py
from __future__ import annotations

from threading import Thread, Lock
from queue import Queue
import socket

from .ListenerSocket import ListenerSocket
from .Packet import Packet
from .PacketBuilder import PacketBuilder


class ListenerThread(Thread):
    """A thread handler for a listener that will continuously listen
    on a specified port and insert all data received into a queue
    """

    __instance: ListenerThread = None

    def __new__(cls, port: int, packet_builder: PacketBuilder) -> None:
        """
        :param port: the port to listen on
        :type port: int
        :param packet_builder: the packet builder to use
        :type packet_builder: PacketBuilder
        """
        if cls.__instance is None:
            cls.__instance = super(ListenerThread, cls).__new__(cls)
            cls.__instance.listener_socket: ListenerSocket = ListenerSocket(
                port, packet_builder
            )
            cls.__instance.packet_queue: Queue = Queue()
            cls.__instance.running = True
            cls.__instance.mutex = Lock()
        return cls.__instance

    def __init__(self, *args, **kwargs) -> None:
        super().__init__()

    def get_instance(cls) -> ListenerThread:
        """Returns the instance of the listener thread

        :return: the instance of the listener thread
        :rtype: ListenerThread
        """
        return cls.__instance

    def run(self) -> None:
        while self.running:
            try:
                new_packet: Packet = self.listener_socket.get_packet()
                self._add_to_queue(new_packet)
            except socket.error:
                self.running = False

    def get_packets(self, num_of_packets: int = None) -> list[Packet]:
        """Retrieves a list of packets from the packet queue.
        Returns all packets if a number wasnt specified.

        :param num_of_packets: the number of packets to retrieve
        :type num_of_packets: int
        :return: a packet
        :rtype: Packet
        """
        if num_of_packets is None:
            num_of_packets = self.packet_queue.qsize()

        with self.mutex:
            packets: list[Packet] = [
                self.packet_queue.get(False) for _ in range(num_of_packets)
            ]

        return packets

    def get_packet(self) -> Packet:
        """Retrieves a packet from the packet queue

        :return: a packet
        :rtype: Packet
        """
        return self.get_packets(1)[0]

    def flush_packets_queue(self) -> None:
        """Flushes the packet queue
        """
        with self.mutex:
            self.packet_queue.queue.clear()

    def is_running(self) -> bool:
        """Returns whether the thread is running

        :return: whether the thread is running
        :rtype: bool
        """
        return self.running

    def _add_to_queue(self, packet: Packet) -> None:
        """Adds a packet to the packet queue

        :param packet: the packet to add
        :type packet: Packet
        """

        with self.mutex:
            if packet.is_single_instance():
                for packet_in_queue in self.packet_queue.queue:
                    if packet_in_queue.header == packet.header:
                        self.packet_queue.queue.remove(packet_in_queue)
                        break
            self.packet_queue.put(packet)

File: RIONet/ga2230net/src/main/java/org/ga2230net/ListenerThread.java
package org.ga2230net;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.locks.ReentrantLock;

/**
 * A thread that listens for incoming packets from a sender and adds them to a queue
 */
public class ListenerThread extends Thread {
    private final Queue<Packet> packetQueue;
    private final ListenerSocket listenerSocket;
    private final ReentrantLock lock;
    private boolean running;
    private static ListenerThread listenerThread = null;

    /**
     * create a new listener thread
     * @param port the port to listen on
     * @param builder the packet builder to use to build packets
     * @throws IOException if an error occurs while creating the listener sockets
     */
    private ListenerThread(int port, PacketBuilder builder) throws IOException {
        lock = new ReentrantLock();
        packetQueue = new LinkedList<>();
        listenerSocket = new ListenerSocket(port, builder);
    }

    /**
     * initializes the single ListenerThread instance
     * @param port the port to listen on
     * @param builder the builder to use for parsing raw packets
     * @throws IOException if an IO error occurred
     */
    public static void initListener(int port, PacketBuilder builder) throws  IOException {
        listenerThread = new ListenerThread(port, builder);
    }

    /**
     * @return the single ListenerThread instance, null if the listener wasn't initialized
     */
    public static ListenerThread getInstance() {
        return listenerThread;
    }

    public void run() {
        running = true;
        while (running) {
            try {
                Packet packet = listenerSocket.getPacket();
                addPacket(packet);
            } catch (IOException e) {
                e.printStackTrace();
                running = false;
            }
        }
    }

    /**
     * get the next n packets from the received packets queue
     *
     * @return a Packet array from the threads packet queue, returns null if
     *         the queue is empty
     */
    public Packet[] getPackets(int numOfPackets) {
        lock.lock();
        try {
            Packet[] packets = new Packet[numOfPackets];
            for(int i = 0; i < numOfPackets; i++){
                packets[i] = packetQueue.poll();
            }
            return packets;
        } finally {
            lock.unlock();
        }
    }

    /**
     * @return an array of all packets left in the packet queue
     */
    public Packet[] getPackets() {
        lock.lock();
        try {
            Packet[] packets = new Packet[packetQueue.size()];
            for(int i = 0; i < packets.length; i++){
                packets[i] = packetQueue.poll();
            }
            return packets;
        } finally {
            lock.unlock();
        }
    }

    /**
     * @return the latest packet recieved in the queue
     */
    public Packet getPacket() {
        lock.lock();
        try {
            return packetQueue.poll();
        } finally {
            lock.unlock();
        }
    }

    /**
     * flushes the packet queue and thus deleting all packets in it
     */
    public void flushPacketsQueue() {
        lock.lock();
        try {
            packetQueue.clear();
        } finally {
            lock.unlock();
        }
    }

    public boolean isRunning() {
        return running;
    }

    private void addPacket(Packet packet) {
        lock.lock();
        try {
            if (packet.isSingleInstance()) {
                for (Packet p: packetQueue) {
                    if (p.getHeader().equals(packet.getHeader())) {
                        packetQueue.remove(p);
                        break;
                    }
                }
            }
            packetQueue.add(packet);
        } finally {
            lock.unlock();
        }
    }
}
